specs/APIReview_AllowExternalDrop.md

```diff
@@ -0,0 +1,100 @@
+# Background
+Currently the functionality of dragging and dropping external objects(e.g. files, hyperlinks) into webview2 is default enabled in webview2 and there is no way to disable it. Some developers may want to disbale this functionality in their applications based upon their scenarios. According to such feature requirements, we add the new API to provide developers with the capability to configure the external drag&drop functionality.
+
+# Description
+We add a new `AllowExternalDrop` property in `CoreWebView2Controller`. 
+This API allows end developers to toggle the external drag&drop functionality easily.
+If it's disabled, drag&drop objects from outside into current webview2 will be forbidden.
+To be more specific, some behaviors listed below will be impacted by the toggle of this property.
+
+* Drag&Drop files on disk to webview2  
+* Drag&Drop hyperlinks from browser to webview2
+* Drag&Drop hyperlinks from one webview2 to another webview2
+
+Some behaviors are not impacted by the toggle of this property like drag&drop selected text to webview2. That means toggle of this property has no impact on the expected behavior of this action.
+
+By default, AllowExternalDrop is enabled to keep consistent with the behavior we had before the API is added.
+
+Please note that drag&drop anything from webview2 to external(outside current webview2) will not be impacted by this property.
+
+# Examples
+## C++
+
+```cpp
+wil::com_ptr<ICoreWebView2Controller4> m_controller4;
+void ToggleAllowDrop()
+{
+    if (m_controller4)
+    {
+        BOOL allowDrop;
+        CHECK_FAILURE(m_controller4->get_AllowDrop(&allowDrop));
+        if (allowDrop)
+        {
+            CHECK_FAILURE(m_controller4->put_AllowDrop(FALSE));
+        }
+        else
+        {
+            CHECK_FAILURE(m_controller4->put_AllowDrop(TRUE));
+        }
+    }
+}
+```
+
+## C#
+```c#
+private CoreWebView2Controller controller
+void ToggleAllowDrop(object target, ExecutedRoutedEventArgs e)
+{
+    if (controller)
+    {
+        controller.AllowDrop = !controller.AllowDrop;
+    }
+}
+```
+
+# Remarks
+The `AllowDrop` property already exists in some controls of .net UI framework like WPF and WinForms. 
+The .net control wrapper for webview2 natively inherits this property. 
+But actually it doesn't really take effect until this new API is added.
```

Comment:
Not sure what this is trying to say. Is it saying that the C# implementation will hook up the WebView2.AllowDrop property to the controller's AllowExternalDrop property?

It looks like the WPF UIElement.AllowDrop property controls both internal and external drop, but this new property controls only external drop. Is it right to connect the two?

Reply:
This text is saying that the UI framework WebView2 control classes (which sort of wrap the CoreWebView2Controller) have properties for disallowing drop inherited from the UI framework, but we previously haven't been able to have them work because there was nothing in CoreWebView2 or CoreWebView2Controller to do this.

If the UI framework's notion of allowing drop doesn't match the new property you're adding we should reconsider. Is the goal to implement the UI framework's allow drop property? If so, can we make your new CoreWebView2Controller.AllowDrop property work in the same way?

Reply:
I removed the text here since it looks kind of confusing in the API review doc. This API review doc should focus on the API itself rather than its wrapper. We can explain to customers when updating webview2 .NET controls. @oldnewthing for your question, I think WPF UIElement AllowDrop means allowing dragging and dropping object from outside the control into the control, I assume this should be external drop which means drop objects from external(outside control) to internal(inside control). I checked the docs for AllowDrop here https://docs.microsoft.com/en-us/dotnet/api/system.windows.uielement.allowdrop?view=net-5.0. But I didn't find any information about the internal drop. If I understand correctly, the AllowDrop property in WPF controls only means allowing drop objects from outside the control into the control. Feel free to correct me if I'm wrong.

Reply:
@oldnewthing @david-risney any insights about this comment?